src/mgd77/mgd77path.1
.TH MGD77PATH 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
mgd77path \- Get full pathname for MGD77 files
.SH SYNOPSIS
\fBmgd77path\fP \fINGDC-ids\fP [ \fB\-D\fP ] [ \fB\-I\fP\fIignore\fP ] [ \fB\-P\fP[\fB-\fP] ] [ \fB\-V\fP ]
.SH DESCRIPTION
\fBmgd77path\fP returns the full pathname to one or more MGD77 files. The pathname returned for a given
cruise may change with time due to reshuffling of disks/subdirectories.
.TP
\fINGDC-ids\fP
Can be one or more of five kinds of specifiers:
.br
1) 8-character NGDC IDs, e.g., 01010083, JA010010etc., etc.
.br
2) 2-character <agency> codes which will return all cruises from each agency.
.br
3) 4-character <agency><vessel> codes, which will return all cruises from those vessels.
.br
4) =<list>, where <list> is a table with NGDC IDs, one per line.
.br
5) If nothing is specified we return all cruises in the data base.
.br
(See mgd77info \fB\-L\fP for agency and vessel codes).
The ".mgd77" or ".nc" extensions will automatically be appended, if needed (use \fB\-I\fP
to ignore certain file types).  Cruise files will be looked for first in the current directory
and second in all directories listed in \fB$MGD77_HOME\fP/mgd77_paths.txt [If \fB$MGD77_HOME\fP
is not set it will default to \fB$GMT_SHAREDIR\fP/mgd77].
.br
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
\fB\-D\fP
Instead of cruise listings, just show the directory paths currently used in the search.
.TP
\fB\-I\fP
Ignore certain data file formats from consideration. Append \fBa|c|m|t\fP to ignore
MGD77 ASCII, MGD77+ netCDF, MGD77T ASCII or plain tab-separated ASCII table files, respectively. The option may
be repeated to ignore more than one format.  [Default ignores none].
.TP
\fB\-P\fP
Display the full path to each cruise [Default].  Optionally, append \fB-\fP which will list
just the cruise IDs instead.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
Reports the total number of cruises found.
.SH EXAMPLES
To obtain pathnames for cruises 01010008 and 01010007, run
.br
.sp
\fBmgd77path\fP 01010008 01010007
.br
.sp
To obtain pathnames for cruises 01010008 and 01010007, but only if there are MGD77+ version in netCDF, run
.br
.sp
\fBmgd77path\fP 01010008 01010007 \fB\-Ia\fP \fB\-It\fP
.br
.sp
To see the list of active directories where MGD77 files might be stored, run
.br
.sp
\fBmgd77path\fP \fB\-D\fP
.SH "SEE ALSO"
.IR GMT (1)
.IR mgd77info (1)
.IR mgd77list (1)
.IR mgd77manage (1)
.IR mgd77track (1)
.SH REFERENCES
Wessel, P., and W. H. F. Smith, 2014, The Generic Mapping Tools (GMT) version
4.5.12 Technical Reference & Cookbook, SOEST/NOAA.
.br
Wessel, P., and W. H. F. Smith, 1998, New, Improved Version of Generic Mapping
Tools Released, EOS Trans., AGU, 79 (47), p. 579.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, EOS Trans., AGU, 76 (33), p. 329.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, http://www.agu.org/eos_elec/95154e.html, Copyright 1995 by the
American Geophysical Union.
.br
Wessel, P., and W. H. F. Smith, 1991, Free Software Helps Map and Display Data,
EOS Trans., AGU, 72 (41), p. 441.
.br
The Marine Geophysical Data Exchange Format - "MGD77", see
\fIhttp://www.ngdc.noaa.gov/mgg/dat/geodas/docs/mgd77.txt\fP
.br


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/mgd77/mgd77path.c
/*--------------------------------------------------------------------
 *	$Id: mgd77path.c 12822 2014-01-31 23:39:56Z remko $
 *
 *    Copyright (c) 2004-2014 by P. Wessel
 *    See README file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * mgd77path accepts MGD77 cruise names and returns the full system
 * path to the file(s).
 *
 * Author:	Paul Wessel
 * Date:	26-AUG-2004
 * Version:	1.0 Based on the old gmtpath.c
 *
 *
 */
 
#define THIS_MODULE_NAME	"mgd77path"
#define THIS_MODULE_LIB		"mgd77"
#define THIS_MODULE_PURPOSE	"Return paths to MGD77 cruises and directories"

#include "gmt_dev.h"
#include "mgd77.h"

#define GMT_PROG_OPTIONS "-V"

struct MGD77PATH_CTRL {	/* All control options for this program (except common args) */
	/* active is true if the option has been activated */
	struct A {	/* -A */
		bool active;
		bool mode;
	} A;
	struct D {	/* -D */
		bool active;
	} D;
	struct I {	/* -I */
		bool active;
		unsigned int n;
		char code[3];
	} I;
};

void *New_mgd77path_Ctrl (struct GMT_CTRL *GMT) {	/* Allocate and initialize a new control structure */
	struct MGD77PATH_CTRL *C = NULL;
	
	C = GMT_memory (GMT, NULL, 1, struct MGD77PATH_CTRL);
	
	/* Initialize values whose defaults are not 0/false/NULL */
	
	return (C);
}

void Free_mgd77path_Ctrl (struct GMT_CTRL *GMT, struct MGD77PATH_CTRL *C) {	/* Deallocate control structure */
	if (!C) return;
	GMT_free (GMT, C);	
}

int GMT_mgd77path_usage (struct GMTAPI_CTRL *API, int level)
{
	GMT_show_name_and_purpose (API, THIS_MODULE_LIB, THIS_MODULE_NAME, THIS_MODULE_PURPOSE);
	if (level == GMT_MODULE_PURPOSE) return (GMT_NOERROR);
	GMT_Message (API, GMT_TIME_NONE, "usage: mgd77path <cruise(s)> A[-] -D [-I<code>] [%s]\n\n", GMT_V_OPT);
        
	if (level == GMT_SYNOPSIS) return (EXIT_FAILURE);
             
	MGD77_Cruise_Explain (API->GMT);
	GMT_Message (API, GMT_TIME_NONE, "\t-A List full cruise pAths [Default].  Append - to only get cruise names.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t-D List all directories with MGD77 files instead.\n");
	GMT_Message (API, GMT_TIME_NONE, "\tOPTIONS:\n\n");
	GMT_Message (API, GMT_TIME_NONE, "\t-I Ignore certain data file formats from consideration. Append combination of act to ignore\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   (a) MGD77 ASCII, (c) MGD77+ netCDF, (m) MGD77T ASCII, or (t) plain table files. [Default ignores none].\n");
	GMT_Option (API, "V,.");
	
	return (EXIT_FAILURE);
}

int GMT_mgd77path_parse (struct GMT_CTRL *GMT, struct MGD77PATH_CTRL *Ctrl, struct GMT_OPTION *options)
{
	/* This parses the options provided to mgd77path and sets parameters in CTRL.
	 * Any GMT common options will override values set previously by other commands.
	 * It also replaces any file names specified as input or output with the data ID
	 * returned when registering these sources/destinations with the API.
	 */

	unsigned int n_errors = 0;
	struct GMT_OPTION *opt = NULL;
	struct GMTAPI_CTRL *API = GMT->parent;

	for (opt = options; opt; opt = opt->next) {
		switch (opt->option) {

			case '<':	/* Skip input files */
			case '#':	/* Skip input files confused as numbers (e.g. 123456) */
				break;

			/* Processes program-specific parameters */

			case 'P':
				if (GMT_compat_check (GMT, 4)) {
					GMT_Report (API, GMT_MSG_COMPAT, "Warning: -P is deprecated; use -A instead mext time.\n");
					Ctrl->A.active = true;
				}
				else {
					n_errors += GMT_default_error (GMT, opt->option);
					break;
				}
			case 'A':	/* Show list of paths to MGD77 files */
				Ctrl->A.active = true;
				if (opt->arg[0] == '-') Ctrl->A.mode = true;
				break;
				
			case 'D':	/* Show list of directories with MGD77 files */
				Ctrl->D.active = true;
				break;

			case 'I':
				Ctrl->I.active = true;
				if (Ctrl->I.n < 3) {
					if (strchr ("act", (int)opt->arg[0]))
						Ctrl->I.code[Ctrl->I.n++] = opt->arg[0];
					else {
						GMT_Report (API, GMT_MSG_NORMAL, "Option -I Bad modifier (%c). Use -Ia|c|t!\n", opt->arg[0]);
						n_errors++;
					}
				}
				else {
					GMT_Report (API, GMT_MSG_NORMAL, "Option -I: Can only be applied 0-2 times\n");
					n_errors++;
				}
				break;

			default:	/* Report bad options */
				n_errors += GMT_default_error (GMT, opt->option);
				break;
		}
	}
	n_errors += GMT_check_condition (GMT, Ctrl->A.active && Ctrl->D.active, "Syntax error: Only one of -A -D may be used\n");

	return (n_errors ? GMT_PARSE_ERROR : GMT_OK);
}

#define bailout(code) {GMT_Free_Options (mode); return (code);}
#define Return(code) {Free_mgd77path_Ctrl (GMT, Ctrl); GMT_end_module (GMT, GMT_cpy); bailout (code);}

int GMT_mgd77path (void *V_API, int mode, void *args)
{
	unsigned int i, n_cruises = 0, n_paths;
	int error = 0;
	
	char path[GMT_BUFSIZ] = {""}, **list = NULL;
	
	struct MGD77_CONTROL M;
	struct MGD77PATH_CTRL *Ctrl = NULL;
	struct GMT_CTRL *GMT = NULL, *GMT_cpy = NULL;
	struct GMT_OPTION *options = NULL;
	struct GMTAPI_CTRL *API = GMT_get_API_ptr (V_API);	/* Cast from void to GMTAPI_CTRL pointer */

	/*----------------------- Standard module initialization and parsing ----------------------*/

	if (API == NULL) return (GMT_NOT_A_SESSION);
	if (mode == GMT_MODULE_PURPOSE) return (GMT_mgd77path_usage (API, GMT_MODULE_PURPOSE));	/* Return the purpose of program */
	options = GMT_Create_Options (API, mode, args);	if (API->error) return (API->error);	/* Set or get option list */

	if (!options || options->option == GMT_OPT_USAGE) bailout (GMT_mgd77path_usage (API, GMT_USAGE));	/* Return the usage message */
	if (options->option == GMT_OPT_SYNOPSIS) bailout (GMT_mgd77path_usage (API, GMT_SYNOPSIS));	/* Return the synopsis */

	/* Parse the command-line arguments */

	GMT = GMT_begin_module (API, THIS_MODULE_LIB, THIS_MODULE_NAME, &GMT_cpy); /* Save current state */
	if (GMT_Parse_Common (API, GMT_PROG_OPTIONS, options)) Return (API->error);
	Ctrl = New_mgd77path_Ctrl (GMT);	/* Allocate and initialize a new control structure */
	if ((error = GMT_mgd77path_parse (GMT, Ctrl, options))) Return (error);
	
	/*---------------------------- This is the mgd77path main code ----------------------------*/

	MGD77_Init (GMT, &M);			/* Initialize MGD77 Machinery */

	if (Ctrl->I.active) MGD77_Process_Ignore (GMT, 'I', Ctrl->I.code);

	if (Ctrl->D.active) {	/* Just list the current active MGD77 data directories and exit */
		printf ("# Currently, your $MGD77_HOME is set to: %s\n", M.MGD77_HOME);
		printf ("# $MGD77_HOME/mgd77_paths.txt contains these directories:\n");
		for (i = 0; i < M.n_MGD77_paths; i++) printf ("%s\n", M.MGD77_datadir[i]);
		Return (EXIT_SUCCESS);
	}

	n_paths = MGD77_Path_Expand (GMT, &M, options, &list);	/* Get list of requested IDs */

	if (n_paths == 0) {
		GMT_Report (API, GMT_MSG_NORMAL, "No cruises found\n");
		Return (EXIT_FAILURE);
	}
	
	for (i = 0; i < n_paths; i++) {		/* Process each ID */
 		if (MGD77_Get_Path (GMT, path, list[i], &M))
			GMT_Report (API, GMT_MSG_NORMAL, "Cannot find cruise %s\n", list[i]);
		else if (Ctrl->A.mode) {
			printf ("%s\n", list[i]);
			n_cruises++;
		}
		else {
			printf ("%s\n", path);
			n_cruises++;
		}
	}
	
	GMT_Report (API, GMT_MSG_VERBOSE, "Found %d cruises\n", n_cruises);
	
	MGD77_Path_Free (GMT, n_paths, list);
	MGD77_end (GMT, &M);
	
	Return (GMT_OK);
}
